controller/raft/handlers.go

```diff
@@ -0,0 +1,225 @@
+	}
+}
+
+type joinHandler struct {
+	controller *Controller
+}
+
+func (self *joinHandler) ContentType() int32 {
+	return JoinRequestType
+}
+
+func (self *joinHandler) HandleReceive(m *channel.Message, ch channel.Channel) {
+	go func() {
+		req := &JoinRequest{}
+		err := req.Decode(m)
+
+		if err != nil {
+			logrus.WithError(err).Error("error decoding join request")
+			sendErrorResponse(m, ch, err, ErrorCodeBadMessage)
+			return
+		}
+
+		logrus.Infof("received join request id: %v, addr: %v, voter: %v", req.Id, req.Addr, !req.ReadOnly)
+
+		err = self.controller.HandleJoin(req)
+		if err != nil {
+			if errors.Is(err, raft.ErrNotLeader) {
+				sendErrorResponse(m, ch, err, ErrorCodeNotLeader)
+			} else {
+				sendErrorResponse(m, ch, err, ErrorCodeGeneric)
+			}
+		} else {
+			resp := channel.NewMessage(SuccessResponseType, nil)
+			resp.ReplyTo(m)
+			if sendErr := ch.Send(resp); sendErr != nil {
+				logrus.WithError(sendErr).Error("error while sending success response")
+			}
+		}
+	}()
+}
+
+func NewRemoveHandler(controller *Controller) channel.TypedReceiveHandler {
+	return &removeHandler{
+		controller: controller,
+	}
+}
+
+type removeHandler struct {
+	controller *Controller
+}
+
+func (self *removeHandler) ContentType() int32 {
+	return RemoveRequestType
+}
+
+func (self *removeHandler) HandleReceive(m *channel.Message, ch channel.Channel) {
+	go func() {
+		req := &RemoveRequest{}
+		err := req.Decode(m)
+
+		if err != nil {
+			logrus.WithError(err).Error("error decoding remove request")
+			sendErrorResponse(m, ch, err, ErrorCodeBadMessage)
+			return
+		}
+
+		logrus.Infof("received remove request id: %v", req.Id)
+
+		err = self.controller.HandleRemove(req)
+		if err != nil {
+			if errors.Is(err, raft.ErrNotLeader) {
+				sendErrorResponse(m, ch, err, ErrorCodeNotLeader)
+			} else {
+				sendErrorResponse(m, ch, err, ErrorCodeGeneric)
+			}
+		} else {
+			resp := channel.NewMessage(SuccessResponseType, nil)
+			resp.ReplyTo(m)
+			if sendErr := ch.Send(resp); sendErr != nil {
+				logrus.WithError(sendErr).Error("error while sending success response")
+			}
+		}
+	}()
+}
+
+func NewCommandHandler(controller *Controller) channel.TypedReceiveHandler {
+	return &commandHandler{
+		controller: controller,
+	}
+}
+
+type commandHandler struct {
+	controller *Controller
+}
+
+func (self *commandHandler) ContentType() int32 {
+	return NewLogEntryType
+}
+
+// possible error types
+// 1. unable to decode command
+// 2. Pre-submit validation failure
+// 3. Raft application failure
+
+// TODO: Use goroutines.Pool
+func (self *commandHandler) HandleReceive(m *channel.Message, ch channel.Channel) {
+	go func() {
+		if err := self.controller.ApplyEncodedCommand(m.Body); err != nil {
+			sendErrorResponseCalculateType(m, ch, err)
+			return
+		}
+		sendSuccessResponse(m, ch)
+	}()
+}
+
+func sendErrorResponseCalculateType(m *channel.Message, ch channel.Channel, err error) {
+	// TODO: Handle more error types explicitly
```

Comment:
Is there a reason to handle more explicitly? Is this just a maintenance block where we have to keep adding more types here?

Reply:
Converted code to use ApiError consistenly when returning errors, so that way we only have api errors and any raft specific errors we may want to handle.